A rewrite of test_translations, adding new features & flexibility

ferdnyc opened the pull request:
I figure if we're gonna do this, let's really do this. The new `test_translations.py` is modular Python containing functions which can be imported, or it can be run as a standalone script. When run as a script, it will exit with a nonzero exit code when errors are encountered. If imported, the `scan_all()` function will raise an exception on error.

Filenames for one or many `.qm` files can be given on the command line in script mode, to check a certain specific file or group of files. If no arguments are given, it falls back to the same method of scanning the directory, then checks all files.

### Errors detected

The error checking is greatly expanded, and now checks all of the following conditions when comparing source strings to translations:
  - The count of `%s`, `%d`, and `%f` escapes is the same for each
  - If any <code>%(<var>var</var>)<var>x</var></code> named placeholders are used:
    - The translation must use all of the same variable names (<code><var>var</var></code>)
    - The type code (<code><var>x</var></code>) for each named variable must be the same

Raw count of `%` characters is NOT used. It's far too inaccurate, both creating false errors _and_ missing real ones. (Like translations containing `% s` instead of `%s` — we have several of those, it turns out.)

- Counts failure to load any translation file as an error, which will result in non-zero exit status

### Example output

When all is well, it just reports success in green, after a summary of the source string and translation file counts (that's always output, for both success and failure):

![image](https://user-images.githubusercontent.com/538020/138364349-91b4bee1-e4d8-4bbf-b504-3858c3c87a81.png)

If there are problems, they're now shown grouped by `.qm` file, with the name of the `.pot` source shown next to each source string, followed by the bad translation. Here are results from a translation file I deliberately inserted errors into:

![image](https://user-images.githubusercontent.com/538020/138364268-4f376e87-df98-40d5-9c2b-4c84566519f2.png)

### `color` utility class
The color-formatting escapes are now provided via a `color` class, which has classmethods (`color.yellow()` and `color.green()` are added to `color.red()`) that will decorate the string representation of its arguments and return an escape-wrapped string. So, for example, you can do this:

![image](https://user-images.githubusercontent.com/538020/138363456-32cca428-c5b4-4580-96b9-0180ca88a81b.png)

ferdnyc committed:
test_translations: rewrite w/ added features

- Modular code can either be imported or run as a script
- One or many .qm filenames can be passed as arguments, to check
  a subset of all files

- Greatly expanded scanning, now checks all of the following
  conditions when comparing source strings to translations:
  - The number of '%s', '%d', and '%f' escapes is the same
  - If any '%(var)x' named placeholders are used:
    - The translation uses all of the same variable names
    - The type code for each named variable is the same
  - Raw count of '%' characters is NOT used (too inaccurate)

- Checks all strings from all .pot sources against each .qm
  in turn, so results are grouped by .qm file
- Counts failure to load any translation file as an error, will
  result in non-zero exit status

- Color-formatting now defined as a class, with methods to wrap
  strings in color styling, e.g.
  `print("Only part of this line is " + color.red("red text"))`
- color.yellow() and color.green() styling added

ferdnyc commented:
And here are real results, run over all of the language files on the `develop` branch:

![image](https://user-images.githubusercontent.com/538020/138364508-f0a74636-c1b5-4f42-835e-f395c7c786fd.png)

So, a few problems we didn't know about, including whatever translator is inserting spaces into the `%` escapes. And one or more who seem to occasionally _add_ a `%` escape that isn't in the source string?

ferdnyc commented:
(The CI failures are a sign the code is _working_, and not passing the translations due to the problems I just pointed out. Which is actually why that job should really be in a separate workflow, come to think of it, rather than part of the main build. (It's also not necessary to run it in each of the matrix builds, the results won't be any different in ubuntu 18.04 vs 20.04.)

ferdnyc commented:
My added commit moves the translation-file scanning into a separate job within the CI workflow, it runs independently and will pass/fail separately from the build jobs (which can in turn pass/fail independent of the translation checks).

edumucelli commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
+TAG_RE = re.compile(r'%\(([^\)]*)\)(.)')
+
+
+class color:
```
This really looks line an Enum to me, a set of name constants, I'd suggest using the std Enum for it.

edumucelli replied:
```suggestion
class Color:
```

ferdnyc replied:
Done, and all of the calls to the classmethods updated accordingly.

ferdnyc replied:
As far as the Enum... I mean, the `Color` class definitely isn't an enum, it's a string-processing class. 

`Color.red(*args)` returns a string that _starts with_ `Color._red`, ends with `Color._reset`, and has the string representation of `*args` in between.

I could make the color escape values a separate Enum of their own, but they're never meant to be accessed outside of the classmethods so is it worth it? It'd just be swapping out a reference to an internal class string member with a reference to a different class's internal string member, arguably at the expense of clarity.

If I'm following your suggestion, something like,
```python3
ColorEscapes = Enum(
    'ColorEscapes', {
        "RED": "\u001b[31m",
        "YELLOW": "\u001b[33m",
        "GREEN": "\u001b[32m",
        "RESET": "\u001b[0m",
    })
```
And then in the `Color` class, calls like...
```python3
@staticmethod
def red(*args):
    return ColorEscapes.RED + str(*args) + ColorEscapes.RESET
```
I'm debating whether that seems more clear or _less_ clear, TBH. And it would make the escapes visible outside the `Color` class (unless I define the enum _inside_ the `Color` class), which I'm not sure is desirable.

edumucelli commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
-if found_errors:
-    raise(Exception("Errors detected during translation testing! See above."))
+
+def build_stringlists() -> dict:
```
I'd suggest using the `typing` package. Although we have discussed and you thoroughly explained you think that is not a priority, but since you are making an effort here, I'd do it using `typing.Dict`, we may profit from your effort here in the future. Actually a `Dict[str, List]` if I got what the method is doing.

ferdnyc replied:
I have typed the module fully, and remembered to run it through `mypy` this time so it's actually _right_ :laughing: It even passes `mypy --disallow-untyped-defs`. Go, me.

edumucelli replied:
Great! I can work on adding the `mypy` to the ci, we can already start adding this module to the list of `opt-in` files, so that we do not accidentally remove/revert the typing effort done here.

edumucelli commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
+        if not errset:
+            continue
+        width = len(pot)
+        msg = "Invalid"
```
I think this `msg` does not need to be reset on every iteration of the loop, I would suggest moving it to the outside.

ferdnyc replied:
Fair point. I suspect Python may optimize such things away when byte-compiling, but there's no sense in making it have to.

(I was originally going to just hardcode it in the f-string, but I don't think there's a way to do that and still match the width of the `pot` string. So then I was going to generate a new `len(pot)`-width version each loop, but I realized that was stupid because <code>{<var>variable</var>:><var>length</variable>}</code> exists. But I apparently stopped one optimization too soon.)

ferdnyc replied:
(Now that I'm rereading it, I recall I also wasn't worried about optimization because this loop only fires once for each _error_ found. Ideally, it'll never run! :grin: )

edumucelli commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
+          + f"{color.yellow(lang_count)} translation files.")
+    if total_errors > 0:
+        msg = f"Found {total_errors} translation errors! See above."
+        raise Exception(msg)
```
Maybe creating an specific Exception to the semantic more clear, e.g., `TranslationError` and raising it here. I am not sure if the codebase has already some policy on defining exceptions.

ferdnyc replied:
A _**policy**_, no, definitely not. But I completely agree re: semantic clarity. So, what the hey? Let's define an exception. There's a first time for everything!

(Sadly true. Looking around, I have a couple of other worktrees with exception definitions in them, but so far nothing submitted. Though I really _should_ submit the one that replaces an abuse of `ModuleNotFoundError`.)

edumucelli replied:
Yes, because at some point (hopefully) the business exceptions are common in multiple parts. They will show similar validation mechanisms that can grow into common code. In this case here I do not now how reusable they will be, so that is just a comment, feel free to dismiss it.

edumucelli reviewed (commented):
I won't approve because I do not feel legitimate to approve anything on the project, but I think this looks neat and improves a lot from the current state.

JacksonRG commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
+    errors.update({
+        s: translations[s]
+        for s, (s_vars, t_vars) in named_variables.items()
+        if sorted(s_vars) != sorted(t_vars)
```
(Admittedly minor nitpick): My only question on this is if a syntactical difference in languages requires a repeat of a variable (probably repeating a name to remove ambiguity, not repeating a number). Would it make sense to compare the sets of variables rather than the lists?

ferdnyc replied:
@JacksonRG Hmm... no, that's a really good point. I'd considered the need to reorder variables (hence the use of `sorted()` when comparing the lists), but not that a variable might be repeated in the source language but not the target (or vice versa, I suppose).

Now that you've got me thinking about it, there's also the possibility that a variable might not make sense in the target language and need to be dropped. It is legal for a `%`-interpolation to _not use_ all of the variables it's passed, so really `t_vars` only needs to be a _strict subset_ of `s_vars` to be valid. 

If I compare using _sets_ I can (a) skip the `sorted()` and (b) make the condition `not set(t_vars) <= set(s_vars)` instead of the `!=` comparison. Ooh, I like that!

JacksonRG replied:
Now you're telling me set overloads the comparison operators? Oh how my list of python tricks grows. Figured it'd have to be `setA.contains(setB)` or similar. That's awesome :sunglasses:

ferdnyc replied:
@JacksonRG 

Oh, sets overload, like, _all_ the operators:
```python3
>>> setA = {"red", "green", "blue"}
>>> setB = {"one", "two", "red", "blue"}
>>> setA | setB  # Union
{'blue', 'one', 'two', 'green', 'red'}
>>> setA - setB  # Difference
{'green'}
>>> setB & setA  # Intersection
{'blue', 'red'}
>>> setB ^ setA  # XOR
{'one', 'two', 'green'}
>>> (setB ^ setA) < setB
False
>>> (setB & setA) < setB
True
# (There are also modifying forms: |=, ^=, -=, &=, etc.)
```
(With weird exceptions. Like, `setA | setB` is valid, but `setA + setB` is a `TypeError`. **#ForSomeReason!**)

JacksonRG commented on src/language/test_translations.py:
```diff
@@ -30,52 +30,153 @@
+            s.count('%s') != t.count('%s'),
+            s.count('%d') != t.count('%d'),
+            s.count('%f') != t.count('%f')])
+    }
```
I like how simple this is to read. Didn't know the { `statement` `for ...` `if ...` } could be used to create dicts and lists! Though I've used it as a combined `map` and `filter` function before.

ferdnyc replied:
@JacksonRG Yeah, [dict comprehensions (PEP 274)](https://www.python.org/dev/peps/pep-0274/) were added relatively "late" to the Python syntax.

(They were added in Python 3.0 & 2.7, released December 2008 & July 2010 respectively. As opposed to list comprehensions, which were defined in [PEP 202](https://www.python.org/dev/peps/pep-0202/) and released as part of Python 2.0 in July 2000. ...I would _love_ it if someone dug up a reference to how they originally got the name "comprehensions", because f--- if I know. Even in PEP 202 the name is unexplained and just taken as a given.)

Since Python 2.2, `dict()` has accepted a list of `(k, v)` tuples, so it's been possible to fake dict comprehensions anyway with `dict([(k, v) for k, v in some_expression_that_yields_k_and_v])`. The dict comprehension syntax just streamlines that for clarity.

Weirdly, while list comprehensions are covered nicely [in the Python Tutorial](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions), dict and set comprehensions get the shaft there.

The thing I like best about them is, because they're totally self-contained, you can use temporary variable names like `s` and `t` without feeling too guilty, since their only uses _have_ to be **very** local. (The temporaries are evaluated in a local scope to prevent their definitions escaping the comprehension, unlike in normal unrolled loops where any variables you define are hoisted to the function level.) Being able to be temporarily concise is nice when your _non_-temporary variable names are long enough to be unambiguous and self-documenting.

Even in its unrolled, brief loop form, this feels like ugly code relative to the comprehension:
```python
errors = {}
for s, t in translations.items():
    if any([s.count('%s') != t.count('%s'),
            s.count('%d') != t.count('%d'),
            s.count('%f') != t.count('%f')
           ]):
       errors[s] = t
```
Not to mention the pointlessness of having to create the empty dict first.

If it's a longer loop that's making more complex decisions about what goes into the dict, then that's probably a sign the contents should be split out into a separate function (that can then be used in a comprehension).

JacksonRG reviewed (approved):
Just 1 question about a very edge case. Approving anyways because I don't think we currently allow that case either.

ferdnyc committed:
Apply review suggestions

Co-authored-by: Eduardo <edumucelli@gmail.com>
Co-authored-by: Jackson <jackson@ganymede.home>

ferdnyc committed:
Move translation tests to conditional workflow

- Will only auto-trigger on changes to src/language/ files
- Can also be manually run by committers

ferdnyc commented:
Rather than having the CI checking (and failing) every PR in the repo whenever there are translation issues, I moved things around again so that the "Validate translations" job only runs when paths under `src/language/` are changed. (So, _this_ PR still triggers it.)

It can also be manually run from the Actions tab, by committers. (At least, _hopefully_ that'll be the case after merge to the default branch, now that I've nailed down the syntax with [some help from the GitHub Community](https://github.community/t/problem-using-workflow-dispatch-trigger-with-other-workflow-triggers/183242).)

jonoomph commented:
@ferdnyc This is awesome! Thanks for digging into this and improving our error detection. I'm working on fixing all these translations now, and then I'll rerun the validate check after merging the fixes.

jonoomph commented:
I've fixed these translations issues in develop, and if we merge develop back into this branch, that test should now pass.

jonoomph commented:
I'll see if I can push to your branch... not sure if that is allowed.

jonoomph commented:
I've decided to just merge this and test locally (since it was taking too long for me to get a local version cloned... not sure why it was going to take hours and hours to clone it... weird)
